Create gss_auth option for paramiko_ssh connection plugin

arnoxit opened the pull request:
##### SUMMARY
Add a new `gss_auth` option for the `paramiko_ssh `connection plugin that will be passed to the upstream `paramiko` library `connect()` method. This will enable GSS-API authentication for paramiko ssh connections, which is required for
Kerberos authentication to succeed.

Set the default for the `gss_auth` option in the `[paramiko_connection]` section to `False` to ensure backwards compatibility.

Fixes #71201

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
paramiko_ssh

##### ADDITIONAL INFORMATION
Without gss_auth enabled for paramiko_ssh against a kerberos authenticated host:
```
(venv) kim.covil@dragon:~/src/ansible$ ansible -m ping -c paramiko -i $HOSTNAME, $HOSTNAME
[WARNING]: You are running the development version of Ansible. You should only run Ansible from "devel" if you are modifying the Ansible engine, or trying out features under development. This is a rapidly changing source of code and can become unstable at any point.
dragon.office.transficc.net | UNREACHABLE! => {
    "changed": false,
    "msg": "Failed to authenticate: Authentication failed.",
    "unreachable": true
}
```

With gss_auth enabled for paramiko_ssh against a kerberos authenticated host:
```
(venv) kim.covil@dragon:~/src/ansible-arnoxit$ cat ansible.cfg
[paramiko_connection]
gss_auth = True
(venv) kim.covil@dragon:~/src/ansible-arnoxit$ ansible -m ping -c paramiko -i $HOSTNAME, $HOSTNAME
[WARNING]: You are running the development version of Ansible. You should only run Ansible from "devel" if you are modifying the Ansible engine, or trying out features under development. This is a rapidly changing source of code and can become unstable at any point.
[WARNING]: Platform linux on host dragon.office.transficc.net is using the discovered Python interpreter at /usr/bin/python, but future installation of another Python interpreter could change the meaning of that path. See https://docs.ansible.com/ansible/devel/reference_appendices/interpreter_discovery.html for
more information.
dragon.office.transficc.net | SUCCESS => {
    "ansible_facts": {
        "discovered_interpreter_python": "/usr/bin/python"
    },
    "changed": false,
    "ping": "pong"
}
```

Commit:
Create gss_auth option for paramiko_ssh connection plugin

- Add a new gss_auth option for the paramiko_ssh connection plugin that will
  be passed to the upstream paramiko library connect method. This will enable
  GSS-API authentication for paramiko ssh connections which is required for
  Kerberos authentication.
- Set the default for the gss_auth option in the paramiko_connection section
  to false to ensure backwards compatibility

Commit:
Omit gss_auth parameter when paramiko version is too old

- Only pass the gss_auth option to paramiko.SSHClient().connect() if the paramiko version
  supports it (paramiko added support for GSS-API in 1.15.0)

bcoca commented on lib/ansible/config/base.yml:
```diff
@@ -1703,6 +1703,14 @@ PARAMIKO_HOST_KEY_AUTO_ADD:
+  env: [{name: ANSIBLE_PARAMIKO_GSS_AUTH}]
+  ini:
+  - {key: gss_auth, section: paramiko_connection}
+  type: boolean
```
this is not needed, it shoudl only be added to the plugin itself

arnoxit replied:
Thanks. I have pushed a change to remove this and the version_added change suggested by @samdoran.

samdoran commented:
This seems like the first of several GSS-API related options that can be added. `paramiko` offers several GSS-API related options that others may need in the future. This opens the door to adding those other options, which we may or may not be ok with.

samdoran commented on lib/ansible/plugins/connection/paramiko_ssh.py:
```diff
@@ -61,6 +61,16 @@
+        ini:
+        - section: paramiko_connection
+          key: gss_auth
+          version_added: '2.9'
```
```suggestion
          version_added: '2.11'
```

Commit:
Remove gss_auth config from base and update ansible version

As per suggestions:
- The PARAMIKO_GSS_AUTH section is not required in config/base.yml
- The ansible version_added should be 2.11

arnoxit commented:
> This seems like the first of several GSS-API related options that can be added. `paramiko` offers several GSS-API related options that others may need in the future. This opens the door to adding those other options, which we may or may not be ok with.

Yes, this is the main option that enables GSS-API authentication in paramiko and, at least for my immediate use case, it was enough to get GSS-API authentication working for paramiko_ssh and therefore also network_cli. 

At the moment paramiko_ssh is not usable for hosts that use kerberos authentication and don't enable passwords or ssh keys (which would be legitimately disabled in a kerberos environment) even though the paramiko library itself does support this. For my use case, the openssh connection plugin does support GSS-API but is not usable with the network_cli connection plugin.

samdoran reviewed (changes requested):
This change looks correct now. Please create a [changelog fragment](https://docs.ansible.com/ansible/latest/community/development_process.html#creating-a-changelog-fragment). See [this fragment](https://github.com/ansible/ansible/blob/c2f528b7686c9573aa8493357219e0bbe4416dd3/changelogs/fragments/reboot-change-default-boot-command.yaml) as an example.

It would be nice to have tests for this. I'm not sure how comfortable you are writing unit tests. We don't have many for this plugin currently, but that's probably the best way to test this since integration tests would require setting up kerberos auth. This is not a blocker, though.

bmillemathias commented on examples/ansible.cfg:
```diff
@@ -365,6 +365,11 @@
+# paramiko will not use GSS-API Authentication unless the gss_auth option
+# is passed. This is required for Kerberos Authentication.
+# Uncomment this line to enable GSS-API Authentication in Paramiko 
+#gss_auth = True
```
Also, should not the commented version be the default value ?

arnoxit replied:
I followed some of the other config examples that also said "Uncomment this line to enable..."

bmillemathias commented on lib/ansible/plugins/connection/paramiko_ssh.py:
```diff
@@ -61,6 +61,16 @@
           - {key: host_key_auto_add, section: paramiko_connection}
         type: boolean
+      gss_auth:
+        default: False
```
Would it not be convenient to have `gss_auth` set by default to True, as using ssh it is already set.

sivel replied:
That would be a behavior change.

Commit:
Add changelogs fragment for #71190/#71201

arnoxit commented:
> This change looks correct now. Please create a [changelog fragment](https://docs.ansible.com/ansible/latest/community/development_process.html#creating-a-changelog-fragment). See [this fragment](https://github.com/ansible/ansible/blob/c2f528b7686c9573aa8493357219e0bbe4416dd3/changelogs/fragments/reboot-change-default-boot-command.yaml) as an example.
> 
> It would be nice to have tests for this. I'm not sure how comfortable you are writing unit tests. We don't have many for this plugin currently, but that's probably the best way to test this since integration tests would require setting up kerberos auth. This is not a blocker, though.

I couldn't see how to set up a test for it without having a kerberos auth environment configured. It might be possible to test that enabling the option means the ssh client attempts gss-api (but then continues to other options). Do you have any pointers for other auth type option tests I could look at?

samdoran commented:
> I couldn't see how to set up a test for it without having a kerberos auth environment configured.

That's why I suggested that unit tests might be easier. Setting up kerberos just for a test is probably quite difficult.

Just run `_connect_uncached()` directly in the unit test and make sure it does the expected thing when it hits the new code path.

You can take a look at `test/units/plugins/connection/test_ssh.py` and add new unit tests to `test/units/plugins/connection/test_paramiko.py`.

mattclay commented:
/azp run

s-hertel commented on lib/ansible/plugins/connection/paramiko_ssh.py:
```diff
@@ -340,6 +350,10 @@ def _connect_uncached(self):
                 ssh_connect_kwargs['auth_timeout'] = self._play_context.timeout
 
+            # paramiko 1.15.0 introduced gss_auth parameter
+            if LooseVersion(paramiko.__version__) >= LooseVersion('1.15.0'):
```
We should give a warning if this is not True and gss_auth is set.

Akasurde commented:
The Paramiko SSH Connection plugin is set for deprecation in 2.21 via https://github.com/ansible/ansible/pull/84727. Thanks.
